File: c/examples/reihenschieber_example.c
/**
 * @file reihenschieber_example.c
 * @brief Reihenschieber Cipher Example
 * 
 * Demonstrates the mechanical polyalphabetic substitution cipher with
 * multiple shift modes, directions, and custom patterns.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../include/cryptology/classical/substitution/polyalphabetic/reihenschieber.h"

int main() {
    printf("=== Reihenschieber Cipher Examples ===\n\n");
    
    char plaintext[] = "HELLO WORLD";
    char key[] = "SECRET";
    char encrypted[100];
    char decrypted[100];
    
    // Basic usage
    printf("1. Basic Encryption/Decryption\n");
    printf("----------------------------------------\n");
    
    if (reihenschieber_encrypt(plaintext, key, NULL, "fixed", "forward", 1, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
        printf("Plaintext:  %s\n", plaintext);
        printf("Key:        %s\n", key);
        printf("Encrypted:  %s\n", encrypted);
        
        if (reihenschieber_decrypt(encrypted, key, NULL, "fixed", "forward", 1, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
            printf("Decrypted:  %s\n", decrypted);
            printf("Success:    %s\n", strcmp(plaintext, decrypted) == 0 ? "Yes" : "No");
        }
    }
    printf("\n");
    
    // Different shift modes
    printf("2. Shift Modes\n");
    printf("----------------------------------------\n");
    
    char test_text[] = "HELLO";
    char test_key[] = "KEY";
    
    // Fixed mode
    printf("Fixed Mode:\n");
    if (reihenschieber_encrypt(test_text, test_key, NULL, "fixed", "forward", 2, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
        printf("  Plaintext:  %s\n", test_text);
        printf("  Encrypted:  %s\n", encrypted);
        
        if (reihenschieber_decrypt(encrypted, test_key, NULL, "fixed", "forward", 2, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
            printf("  Decrypted:  %s\n", decrypted);
        }
    }
    printf("\n");
    
    // Progressive mode
    printf("Progressive Mode:\n");
    if (reihenschieber_encrypt(test_text, test_key, NULL, "progressive", "forward", 1, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
        printf("  Plaintext:  %s\n", test_text);
        printf("  Encrypted:  %s\n", encrypted);
        
        if (reihenschieber_decrypt(encrypted, test_key, NULL, "progressive", "forward", 1, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
            printf("  Decrypted:  %s\n", decrypted);
        }
    }
    printf("\n");
    
    // Custom mode
    printf("Custom Mode:\n");
    int custom_shifts[] = {1, -1, 2, -2, 0};
    if (reihenschieber_encrypt(test_text, test_key, NULL, "custom", "forward", 0, custom_shifts, 5, encrypted, sizeof(encrypted)) == 0) {
        printf("  Plaintext:  %s\n", test_text);
        printf("  Encrypted:  %s\n", encrypted);
        
        if (reihenschieber_decrypt(encrypted, test_key, NULL, "custom", "forward", 0, custom_shifts, 5, decrypted, sizeof(decrypted)) == 0) {
            printf("  Decrypted:  %s\n", decrypted);
        }
    }
    printf("\n");
    
    // Shift directions
    printf("3. Shift Directions\n");
    printf("----------------------------------------\n");
    
    // Forward (default)
    printf("Forward Direction:\n");
    if (reihenschieber_encrypt(test_text, test_key, NULL, "fixed", "forward", 2, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
        printf("  Plaintext:  %s\n", test_text);
        printf("  Encrypted:  %s\n", encrypted);
        
        if (reihenschieber_decrypt(encrypted, test_key, NULL, "fixed", "forward", 2, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
            printf("  Decrypted:  %s\n", decrypted);
        }
    }
    printf("\n");
    
    // Backward
    printf("Backward Direction:\n");
    if (reihenschieber_encrypt(test_text, test_key, NULL, "fixed", "backward", 2, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
        printf("  Plaintext:  %s\n", test_text);
        printf("  Encrypted:  %s\n", encrypted);
        
        if (reihenschieber_decrypt(encrypted, test_key, NULL, "fixed", "backward", 2, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
            printf("  Decrypted:  %s\n", decrypted);
        }
    }
    printf("\n");
    
    // Custom shift patterns
    printf("4. Custom Shift Patterns\n");
    printf("----------------------------------------\n");
    
    int shifts[10];
    char generated_key[20];
    
    // Alternating pattern
    printf("Alternating Pattern:\n");
    if (reihenschieber_produce_custom_shifts("alternating", 5, shifts, sizeof(shifts)) == 0) {
        printf("  Shifts: [");
        for (int i = 0; i < 5; i++) {
            printf("%d", shifts[i]);
            if (i < 4) printf(", ");
        }
        printf("]\n");
        
        if (reihenschieber_encrypt(test_text, test_key, NULL, "custom", "forward", 0, shifts, 5, encrypted, sizeof(encrypted)) == 0) {
            printf("  Plaintext:  %s\n", test_text);
            printf("  Encrypted:  %s\n", encrypted);
            
            if (reihenschieber_decrypt(encrypted, test_key, NULL, "custom", "forward", 0, shifts, 5, decrypted, sizeof(decrypted)) == 0) {
                printf("  Decrypted:  %s\n", decrypted);
            }
        }
    }
    printf("\n");
    
    // Fibonacci pattern
    printf("Fibonacci Pattern:\n");
    if (reihenschieber_produce_custom_shifts("fibonacci", 5, shifts, sizeof(shifts)) == 0) {
        printf("  Shifts: [");
        for (int i = 0; i < 5; i++) {
            printf("%d", shifts[i]);
            if (i < 4) printf(", ");
        }
        printf("]\n");
        
        if (reihenschieber_encrypt(test_text, test_key, NULL, "custom", "forward", 0, shifts, 5, encrypted, sizeof(encrypted)) == 0) {
            printf("  Plaintext:  %s\n", test_text);
            printf("  Encrypted:  %s\n", encrypted);
            
            if (reihenschieber_decrypt(encrypted, test_key, NULL, "custom", "forward", 0, shifts, 5, decrypted, sizeof(decrypted)) == 0) {
                printf("  Decrypted:  %s\n", decrypted);
            }
        }
    }
    printf("\n");
    
    // Prime pattern
    printf("Prime Pattern:\n");
    if (reihenschieber_produce_custom_shifts("prime", 5, shifts, sizeof(shifts)) == 0) {
        printf("  Shifts: [");
        for (int i = 0; i < 5; i++) {
            printf("%d", shifts[i]);
            if (i < 4) printf(", ");
        }
        printf("]\n");
        
        if (reihenschieber_encrypt(test_text, test_key, NULL, "custom", "forward", 0, shifts, 5, encrypted, sizeof(encrypted)) == 0) {
            printf("  Plaintext:  %s\n", test_text);
            printf("  Encrypted:  %s\n", encrypted);
            
            if (reihenschieber_decrypt(encrypted, test_key, NULL, "custom", "forward", 0, shifts, 5, decrypted, sizeof(decrypted)) == 0) {
                printf("  Decrypted:  %s\n", decrypted);
            }
        }
    }
    printf("\n");
    
    // Random key generation
    printf("5. Random Key Generation\n");
    printf("----------------------------------------\n");
    
    char random_text[] = "HELLO WORLD";
    
    // Generate random key
    if (reihenschieber_generate_random_key(5, generated_key, sizeof(generated_key)) == 0) {
        printf("Random Key: %s\n", generated_key);
        
        if (reihenschieber_encrypt(random_text, generated_key, NULL, "fixed", "forward", 1, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
            printf("Plaintext:  %s\n", random_text);
            printf("Encrypted:  %s\n", encrypted);
            
            if (reihenschieber_decrypt(encrypted, generated_key, NULL, "fixed", "forward", 1, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
                printf("Decrypted:  %s\n", decrypted);
            }
        }
    }
    printf("\n");
    
    // Encrypt with random key
    printf("Auto-generated Key:\n");
    if (reihenschieber_encrypt_with_random_key(random_text, 0, encrypted, sizeof(encrypted), generated_key, sizeof(generated_key)) == 0) {
        printf("Auto-generated Key: %s\n", generated_key);
        printf("Plaintext:  %s\n", random_text);
        printf("Encrypted:  %s\n", encrypted);
        
        if (reihenschieber_decrypt(encrypted, generated_key, NULL, "fixed", "forward", 1, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
            printf("Decrypted:  %s\n", decrypted);
        }
    }
    printf("\n");
    
    // Turkish alphabet support
    printf("6. Turkish Alphabet Support\n");
    printf("----------------------------------------\n");
    char turkish_text[] = "MERHABA DUNYA";
    char turkish_key[] = "ANAHTAR";
    
    if (reihenschieber_encrypt_turkish(turkish_text, turkish_key, "fixed", "forward", 1, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
        printf("Turkish Text: %s\n", turkish_text);
        printf("Turkish Key:  %s\n", turkish_key);
        printf("Encrypted:    %s\n", encrypted);
        
        if (reihenschieber_decrypt_turkish(encrypted, turkish_key, "fixed", "forward", 1, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
            printf("Decrypted:    %s\n", decrypted);
            printf("Success:      %s\n", strcmp(turkish_text, decrypted) == 0 ? "Yes" : "No");
        }
    }
    printf("\n");
    
    // Long text example
    printf("7. Long Text Example\n");
    printf("----------------------------------------\n");
    char long_text[] = "THE QUICK BROWN FOX JUMPS OVER THE LAZY DOG";
    char long_key[] = "SECRETKEY";
    
    if (reihenschieber_encrypt(long_text, long_key, NULL, "progressive", "forward", 1, NULL, 0, encrypted, sizeof(encrypted)) == 0) {
        printf("Long Text:   %s\n", long_text);
        printf("Long Key:    %s\n", long_key);
        printf("Encrypted:   %s\n", encrypted);
        
        if (reihenschieber_decrypt(encrypted, long_key, NULL, "progressive", "forward", 1, NULL, 0, decrypted, sizeof(decrypted)) == 0) {
            printf("Decrypted:   %s\n", decrypted);
            printf("Success:     %s\n", strcmp(long_text, decrypted) == 0 ? "Yes" : "No");
        }
    }
    printf("\n");
    
    // Error handling
    printf("8. Error Handling\n");
    printf("----------------------------------------\n");
    
    // Test invalid parameters
    if (reihenschieber_encrypt(NULL, "KEY", NULL, "fixed", "forward", 1, NULL, 0, encrypted, sizeof(encrypted)) == -1) {
        printf("✓ Caught expected error: NULL plaintext\n");
    }
    
    if (reihenschieber_encrypt("HELLO", NULL, NULL, "fixed", "forward", 1, NULL, 0, encrypted, sizeof(encrypted)) == -1) {
        printf("✓ Caught expected error: NULL key\n");
    }
    
    if (reihenschieber_generate_random_key(-1, generated_key, sizeof(generated_key)) == -1) {
        printf("✓ Caught expected error: negative key length\n");
    }
    
    printf("\n=== All Examples Completed Successfully! ===\n");
    
    return 0;
}

File: python/examples/reihenschieber_example.py
"""
Reihenschieber Cipher Examples

Demonstrates the mechanical polyalphabetic substitution cipher with
multiple shift modes, directions, and custom patterns.
"""

from cryptology.classical.substitution.polyalphabetic.reihenschieber import (
    reihenschieber_encrypt,
    reihenschieber_decrypt,
    reihenschieber_generate_random_key,
    reihenschieber_generate_key_for_text,
    reihenschieber_encrypt_with_random_key,
    reihenschieber_produce_custom_shifts,
    reihenschieber_encrypt_turkish,
    reihenschieber_decrypt_turkish,
    TURKISH_ALPHABET
)


def main():
    print("=== Reihenschieber Cipher Examples ===\n")
    
    # Basic usage
    print("1. Basic Encryption/Decryption")
    print("-" * 40)
    plaintext = "HELLO WORLD"
    key = "SECRET"
    
    encrypted = reihenschieber_encrypt(plaintext, key)
    decrypted = reihenschieber_decrypt(encrypted, key)
    
    print(f"Plaintext:  {plaintext}")
    print(f"Key:        {key}")
    print(f"Encrypted:  {encrypted}")
    print(f"Decrypted:  {decrypted}")
    print(f"Success:    {plaintext == decrypted}")
    print()
    
    # Different shift modes
    print("2. Shift Modes")
    print("-" * 40)
    plaintext = "HELLO"
    key = "KEY"
    
    # Fixed mode
    encrypted_fixed = reihenschieber_encrypt(plaintext, key, shift_mode="fixed", shift_amount=2)
    decrypted_fixed = reihenschieber_decrypt(encrypted_fixed, key, shift_mode="fixed", shift_amount=2)
    print(f"Fixed Mode:")
    print(f"  Plaintext:  {plaintext}")
    print(f"  Encrypted:  {encrypted_fixed}")
    print(f"  Decrypted:  {decrypted_fixed}")
    print()
    
    # Progressive mode
    encrypted_progressive = reihenschieber_encrypt(plaintext, key, shift_mode="progressive", shift_amount=1)
    decrypted_progressive = reihenschieber_decrypt(encrypted_progressive, key, shift_mode="progressive", shift_amount=1)
    print(f"Progressive Mode:")
    print(f"  Plaintext:  {plaintext}")
    print(f"  Encrypted:  {encrypted_progressive}")
    print(f"  Decrypted:  {decrypted_progressive}")
    print()
    
    # Custom mode
    custom_shifts = [1, -1, 2, -2, 0]
    encrypted_custom = reihenschieber_encrypt(plaintext, key, shift_mode="custom", custom_shifts=custom_shifts)
    decrypted_custom = reihenschieber_decrypt(encrypted_custom, key, shift_mode="custom", custom_shifts=custom_shifts)
    print(f"Custom Mode:")
    print(f"  Plaintext:  {plaintext}")
    print(f"  Encrypted:  {encrypted_custom}")
    print(f"  Decrypted:  {decrypted_custom}")
    print()
    
    # Shift directions
    print("3. Shift Directions")
    print("-" * 40)
    plaintext = "HELLO"
    key = "KEY"
    
    # Forward (default)
    encrypted_forward = reihenschieber_encrypt(plaintext, key, shift_direction="forward", shift_amount=2)
    decrypted_forward = reihenschieber_decrypt(encrypted_forward, key, shift_direction="forward", shift_amount=2)
    print(f"Forward Direction:")
    print(f"  Plaintext:  {plaintext}")
    print(f"  Encrypted:  {encrypted_forward}")
    print(f"  Decrypted:  {decrypted_forward}")
    print()
    
    # Backward
    encrypted_backward = reihenschieber_encrypt(plaintext, key, shift_direction="backward", shift_amount=2)
    decrypted_backward = reihenschieber_decrypt(encrypted_backward, key, shift_direction="backward", shift_amount=2)
    print(f"Backward Direction:")
    print(f"  Plaintext:  {plaintext}")
    print(f"  Encrypted:  {encrypted_backward}")
    print(f"  Decrypted:  {decrypted_backward}")
    print()
    
    # Custom shift patterns
    print("4. Custom Shift Patterns")
    print("-" * 40)
    plaintext = "HELLO"
    key = "KEY"
    
    # Alternating pattern
    alternating_shifts = reihenschieber_produce_custom_shifts("alternating", 5)
    encrypted_alt = reihenschieber_encrypt(plaintext, key, shift_mode="custom", custom_shifts=alternating_shifts)
    decrypted_alt = reihenschieber_decrypt(encrypted_alt, key, shift_mode="custom", custom_shifts=alternating_shifts)
    print(f"Alternating Pattern: {alternating_shifts}")
    print(f"  Plaintext:  {plaintext}")
    print(f"  Encrypted:  {encrypted_alt}")
    print(f"  Decrypted:  {decrypted_alt}")
    print()
    
    # Fibonacci pattern
    fibonacci_shifts = reihenschieber_produce_custom_shifts("fibonacci", 5)
    encrypted_fib = reihenschieber_encrypt(plaintext, key, shift_mode="custom", custom_shifts=fibonacci_shifts)
    decrypted_fib = reihenschieber_decrypt(encrypted_fib, key, shift_mode="custom", custom_shifts=fibonacci_shifts)
    print(f"Fibonacci Pattern: {fibonacci_shifts}")
    print(f"  Plaintext:  {plaintext}")
    print(f"  Encrypted:  {encrypted_fib}")
    print(f"  Decrypted:  {decrypted_fib}")
    print()
    
    # Prime pattern
    prime_shifts = reihenschieber_produce_custom_shifts("prime", 5)
    encrypted_prime = reihenschieber_encrypt(plaintext, key, shift_mode="custom", custom_shifts=prime_shifts)
    decrypted_prime = reihenschieber_decrypt(encrypted_prime, key, shift_mode="custom", custom_shifts=prime_shifts)
    print(f"Prime Pattern: {prime_shifts}")
    print(f"  Plaintext:  {plaintext}")
    print(f"  Encrypted:  {encrypted_prime}")
    print(f"  Decrypted:  {decrypted_prime}")
    print()
    
    # Random key generation
    print("5. Random Key Generation")
    print("-" * 40)
    plaintext = "HELLO WORLD"
    
    # Generate random key
    random_key = reihenschieber_generate_random_key(5)
    encrypted_random = reihenschieber_encrypt(plaintext, random_key)
    decrypted_random = reihenschieber_decrypt(encrypted_random, random_key)
    
    print(f"Random Key: {random_key}")
    print(f"Plaintext:  {plaintext}")
    print(f"Encrypted:  {encrypted_random}")
    print(f"Decrypted:  {decrypted_random}")
    print()
    
    # Encrypt with random key
    encrypted_auto, auto_key = reihenschieber_encrypt_with_random_key(plaintext)
    decrypted_auto = reihenschieber_decrypt(encrypted_auto, auto_key)
    
    print(f"Auto-generated Key: {auto_key}")
    print(f"Plaintext:  {plaintext}")
    print(f"Encrypted:  {encrypted_auto}")
    print(f"Decrypted:  {decrypted_auto}")
    print()
    
    # Turkish alphabet support
    print("6. Turkish Alphabet Support")
    print("-" * 40)
    turkish_text = "MERHABA DÜNYA"
    turkish_key = "ANAHTAR"
    
    encrypted_turkish = reihenschieber_encrypt_turkish(turkish_text, turkish_key)
    decrypted_turkish = reihenschieber_decrypt_turkish(encrypted_turkish, turkish_key)
    
    print(f"Turkish Text: {turkish_text}")
    print(f"Turkish Key:  {turkish_key}")
    print(f"Encrypted:    {encrypted_turkish}")
    print(f"Decrypted:    {decrypted_turkish}")
    print(f"Success:      {turkish_text == decrypted_turkish}")
    print()
    
    # Long text example
    print("7. Long Text Example")
    print("-" * 40)
    long_text = "THE QUICK BROWN FOX JUMPS OVER THE LAZY DOG"
    long_key = "SECRETKEY"
    
    encrypted_long = reihenschieber_encrypt(long_text, long_key, shift_mode="progressive", shift_amount=1)
    decrypted_long = reihenschieber_decrypt(encrypted_long, long_key, shift_mode="progressive", shift_amount=1)
    
    print(f"Long Text:   {long_text}")
    print(f"Long Key:    {long_key}")
    print(f"Encrypted:   {encrypted_long}")
    print(f"Decrypted:   {decrypted_long}")
    print(f"Success:     {long_text == decrypted_long}")
    print()
    
    # Error handling
    print("8. Error Handling")
    print("-" * 40)
    
    try:
        reihenschieber_encrypt("HELLO123", "KEY")
        print("ERROR: Should have raised ValueError for invalid characters")
    except ValueError as e:
        print(f"✓ Caught expected error: {e}")
    
    try:
        reihenschieber_encrypt("HELLO", "KEY", shift_mode="invalid")
        print("ERROR: Should have raised ValueError for invalid shift mode")
    except ValueError as e:
        print(f"✓ Caught expected error: {e}")
    
    try:
        reihenschieber_generate_random_key(-1)
        print("ERROR: Should have raised ValueError for negative key length")
    except ValueError as e:
        print(f"✓ Caught expected error: {e}")
    
    print("\n=== All Examples Completed Successfully! ===")


if __name__ == "__main__":
    main()
